images/bridge/cmd/bridge/main.go

```diff
@@ -143,99 +120,232 @@ func main() {
+			URL:     mustGetenv("SUBSTRATE_EVENT_COMMANDS_URL"),
+			Command: "events:write",
+		},
+		RequestBodyLogger{},
 	)
 }
 
-type EventCursor struct {
-	LastProcessed event.ID `json:"last_processed"`
+type WriteEventsInput struct {
+	Events []event.PendingEvent `json:"events"`
+}
+
+type WriteEventsReturns struct {
+	IDs []event.ID `json:"ids"`
+
+	FieldsSHA256s []*event.SHA256Digest `json:"fields_sha256s"`
+	DataSHA256s   []*event.SHA256Digest `json:"data_sha256s"`
 }
 
-type EventStreamer struct {
-	NotifyQueue     *notify.Queue
-	CursorNotifiers []notify.Notifier[EventCursor]
+type EventWriteCommand = calls.CommandCall[WriteEventsInput, WriteEventsReturns]
+
+type EventCommands struct {
 	Session         *tracks.Session
-	Subscription    interface{ Serve(context.Context) }
+	EventPathPrefix string
+	BridgeURL       string
+	Handlers        []interface {
+		HandleEvent2(context.Context, tracks.Event) ([]tracks.PathEvent, error)
+	}
+	WriteEvents *EventWriteCommand
 }
 
-func NewEventStreamer(streamURL string, streamQuery *event.Query) *EventStreamer {
-	es := &EventStreamer{}
-	js := httpevents.NewJSONSubscription(
-		"POST",
-		streamURL,
-		streamQuery,
-		notify.On(es.HandleEvent),
-	)
-	js.Initialize()
-	es.Subscription = js
-	return es
+type EventResult struct {
+	Next []event.PendingEvent `json:"next" doc:""`
 }
 
-func (es *EventStreamer) Serve(ctx context.Context) {
-	es.Subscription.Serve(ctx)
+type CommandRuleInput struct {
+	Path     string `json:"path"`
+	Disabled bool   `json:"disabled,omitempty"`
+	Deleted  bool   `json:"deleted,omitempty"`
+
+	Conditions []*event.Query `json:"conditions"`
+	Command    commands.Msg   `json:"command"`
+
+	// Cursor *CommandRuleCursor `json:"-"`
+}
+
+func ptr[T any](t T) *T {
+	return &t
 }
 
-func (es *EventStreamer) HandleEvent(ctx context.Context, n event.Notification, t *struct{}) {
-	events, err := event.Unmarshal[tracks.JSONEvent](n.Events, true)
+func (es *EventCommands) Initialize() {
+	rules := []CommandRuleInput{
+		{
+			Path: "/rules/defs" + es.EventPathPrefix,
+			Conditions: []*event.Query{
+				{
+					EventsWherePrefix: map[string][]event.WherePrefix{
+						"path": {{Prefix: es.EventPathPrefix}},
+					},
+				},
+			},
+			Command: commands.Msg{
+				Meta: commands.Meta{
+					"#/data/parameters/events": {Type: "any"},
+					"#/data/returns/events":    {Type: "any"},
+				},
+				MsgIn: commands.Bindings{
+					"#/msg/data/parameters/events": "#/data/parameters/events",
+				},
+				MsgOut: commands.Bindings{
+					"#/data/returns/next": "#/msg/data/returns/next",
+				},
+				Msg: &commands.Msg{
+					Cap: ptr("reflect"),
+					Data: commands.Fields{
+						"url":  es.BridgeURL,
+						"name": "events:handle",
+					},
+				},
+			},
+		},
+	}
+	events := make([]event.PendingEvent, 0, len(rules))
+	for _, r := range rules {
+		b, err := json.Marshal(r)
+		fatal(err)
+		events = append(events, event.PendingEvent{
+			ConflictKeys: []string{"conditions", "command"},
```

Comment:
@ajbouh I added this, but it seems like a new event shows up each time bridge restarts. Not sure if there's a different format the keys should take here.